### DISCUSSION TITLE
Question, custom decorator

### DISCUSSION THREAD

**Discussion by Amirkaaa** [ag2ai/faststream#810] [Q&A] [answered]
Hi, I just have a question, currently working with faststream and it's great.
I need to write something like custom decorator which will send data to cache, all router functions returns response to cache function, and I don't know how it's possible in the shortest way. 

```
@router.subscriber('topic-name')
@cache_response
def topic_endpoint(data: Any, message: KafkaMessage):
    return { 'modified_data': modified_data, 'correlation_id': message.correlation_id }
```

I've tried many times to write my custom decorator, but seems like I do something wrong or don't understand something, can you help me with one example. Also I've tried to use Middlewares but it didn't helped me.

**Comment by Lancetnik** [comment, marked as answer]
@Amirkaaa hi! Thanks for being with the project

To create custom decorator you just need to create the regular one with the `functools.wraps` using to provide FastStream information about the main handler arguments tupes

This is the correct example

```python
from functools import wraps

from faststream import FastStream, Logger
from faststream.nats import NatsBroker


def decorator(func):
    @wraps(func)
    async def wrapper(*args, **kwargs):
        return await func(*args, **kwargs)
    return wrapper


broker = NatsBroker()
app = FastStream(broker)


@broker.subscriber("test")
@decorator
async def handler(msg: str, logger: Logger):
    logger.info(msg)


@app.after_startup
async def test():
    await broker.publish("Hi!", "test")
```

**Reply by Lancetnik** [reply]
@Amirkaaa can you mark the answer as correct if it helped you. please?
